cpp/tests/rolling/offset_row_window_test.cpp

```diff
@@ -0,0 +1,294 @@
+
+#include <cudf/aggregation.hpp>
+#include <cudf/groupby.hpp>
+#include <cudf/lists/explode.hpp>
+#include <cudf/rolling.hpp>
+#include <cudf/utilities/default_stream.hpp>
+
+template <typename T>
+using fwcw = cudf::test::fixed_width_column_wrapper<T>;
+template <typename T>
+using decimals_column = cudf::test::fixed_point_column_wrapper<T>;
+using ints_column     = fwcw<int32_t>;
+using bigints_column  = fwcw<int64_t>;
+using strings_column  = cudf::test::strings_column_wrapper;
+using lists_column    = cudf::test::lists_column_wrapper<int32_t>;
+using column_ptr      = std::unique_ptr<cudf::column>;
+using cudf::test::iterators::all_nulls;
+using cudf::test::iterators::no_nulls;
+using cudf::test::iterators::nulls_at;
+
+auto constexpr null = int32_t{0};  // NULL representation for int32_t;
+
+struct OffsetRowWindowTest : public cudf::test::BaseFixture {
+  static ints_column const _keys;    // {0, 0, 0, 0, 0, 0, 1, 1, 1, 1};
+  static ints_column const _values;  // {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+
+  struct rolling_runner {
+    cudf::window_bounds _preceding, _following;
+    cudf::size_type _min_periods;
+    bool _grouped = true;
+
+    rolling_runner(cudf::window_bounds const& preceding,
+                   cudf::window_bounds const& following,
+                   cudf::size_type min_periods_ = 1)
+      : _preceding{preceding}, _following{following}, _min_periods{min_periods_}
+    {
+    }
+
+    rolling_runner& min_periods(cudf::size_type min_periods_)
+    {
+      _min_periods = min_periods_;
+      return *this;
+    }
+
+    rolling_runner& grouped(bool grouped_)
+    {
+      _grouped = grouped_;
+      return *this;
+    }
+
+    std::unique_ptr<cudf::column> operator()(cudf::rolling_aggregation const& agg) const
+    {
+      auto const grouping_keys =
+        _grouped ? std::vector<cudf::column_view>{_keys} : std::vector<cudf::column_view>{};
+      return cudf::grouped_rolling_window(
+        cudf::table_view{grouping_keys}, _values, _preceding, _following, _min_periods, agg);
+    }
+  };
+};
+
+ints_column const OffsetRowWindowTest::_keys{0, 0, 0, 0, 0, 0, 1, 1, 1, 1};
+ints_column const OffsetRowWindowTest::_values{0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+
+auto const AGG_COUNT_NON_NULL =
+  cudf::make_count_aggregation<cudf::rolling_aggregation>(cudf::null_policy::EXCLUDE);
+auto const AGG_COUNT_ALL =
+  cudf::make_count_aggregation<cudf::rolling_aggregation>(cudf::null_policy::INCLUDE);
+auto const AGG_MIN          = cudf::make_min_aggregation<cudf::rolling_aggregation>();
+auto const AGG_MAX          = cudf::make_max_aggregation<cudf::rolling_aggregation>();
+auto const AGG_SUM          = cudf::make_sum_aggregation<cudf::rolling_aggregation>();
+auto const AGG_COLLECT_LIST = cudf::make_collect_list_aggregation<cudf::rolling_aggregation>();
+
+TEST_F(OffsetRowWindowTest, OffsetRowWindow_Grouped_3_to_Minus_1)
+{
+  auto const preceding = cudf::window_bounds::get(3);
+  auto const following = cudf::window_bounds::get(-1);
```

Comment:
If I am understanding `following` correctly it could be -4 to capture 4 rows before the current row. We shouldn't have all the 'before' tests use same `following` value.

Reply:
Added better tests for different following values.